### PR TITLE
Implement netty client using the new micronaut-core RawHttpClient

### PR DESCRIPTION
**Pull request by yawkat**
This PR replaces most of the netty-based HTTP client that previously used the low-level micronaut-http-client ConnectionManager API with an implementation based on the new micronaut-http-client-core RawHttpClient introduced in micronaut-core 4.7.0. This offers some advantages:

- Much simpler implementation. Once the legacy implementation is not needed anymore, all classes deprecated in this PR can be removed.
- Possibility to work with non-netty RawHttpClient implementations, once those exist. In particular, this will allow using the JDK http client instead.
- RawHttpClient runs normal micronaut-core ClientFilters, so we won't need the `OciNettyClientFilter` API anymore.

While in theory the new implementation should be a drop-in replacement, it is possible that RawHttpClient differs slightly in behavior. For that reason, I've kept the old implementation around. It can be enabled by the `oci.netty.legacy-netty-client` config property, or the `io.micronaut.oraclecloud.httpclient.netty.legacy-netty-client` system property.

This also means that MicronautHttpRequest and MicronautHttpResponse are actually mostly copied from NettyHttpRequest and NettyHttpResponse. Please consider that when reviewing, the actual changes are not that big.

This PR works and is ready for review, but still needs a release of micronaut-core 4.7.0 (and of micronaut-serialization due to the unrelated https://github.com/micronaut-projects/micronaut-serialization/pull/943 ).

### CODE REVIEW

File: oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/DecidedBodyHandler.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
  * consume the body. First any data buffered by {@link UndecidedBodyHandler} is processed, then a {@link HandlerImpl}
  * is added to the pipeline to directly process the remaining chunks.
  */
+@Deprecated
```
Comment (graemerocher) [inline oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/DecidedBodyHandler.java:35]: use `forRemove=true` on all `@Deprecated` types

File: oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/LimitedBufferingSubscriber.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * {@link Subscriber} implementation that buffers its input up to a certain number of bytes.
+ *
+ * @author Jonas Konrad
+ * @since 4.3
+ */
+final class LimitedBufferingSubscriber implements Subscriber<ByteBuffer<?>>, Closeable {
```
Comment (graemerocher) [inline oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/LimitedBufferingSubscriber.java:36]: add `@Internal`

File: oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * {@link java.util.Map} wrapper around micronaut {@link HttpHeaders}.
+ */
+final class MicronautHeaderMap extends AbstractMap<String, List<String>> {
```
Comment (graemerocher) [inline oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java:32]: add `@Internal`

File: oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            return null;
+        }
+        List<String> items = headers.getAll(s);
+        ((MutableHttpHeaders) headers).remove(s);
+        return items.isEmpty() ? null : items; // follow remove() contract
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+        return key instanceof String && headers.contains((String) key);
```
Comment (graemerocher) [inline oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java]: could probably do `instanceof CharSequence` and then `toString()`

File: oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @Override
+    public Set<String> keySet() {
+        return new KeySet();
+    }
+
+    private class KeySet extends AbstractSet<String> {
```
Comment (graemerocher) [inline oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java]: declare `final`

File: oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            return headers.names().iterator();
+        }
+
+        @Override
+        public int size() {
+            return headers.names().size();
+        }
+    }
+
+    private static class HeaderIterator implements Iterator<Entry<String, List<String>>> {
```
Comment (graemerocher) [inline oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java]: declare `final`

File: oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHttpRequest.java [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+
+import static io.micronaut.oraclecloud.httpclient.netty.NettyClientProperties.CLASS_AND_METHOD_KEY_NAME;
+
+final class MicronautHttpRequest implements HttpRequest {
```
Comment (graemerocher) [inline oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHttpRequest.java:57]: annotate `@Internal`

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java (hunk 4 above)
Comment (yawkat) [reply oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHeaderMap.java]: the java-sdk client API only allows String so there's no point.

File: oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/DecidedBodyHandler.java (hunk 1 above)
Comment (yawkat) [reply oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/DecidedBodyHandler.java:35]: I tried but it marks all uses of the type as an error in IDEA which is a bit confusing. And they're all package-private types anyway so it doesn't matter.

**Review by graemerocher — APPROVED** [review]

**Pull request merged**